### PR TITLE
suggest tuple struct syntax

### PR DESCRIPTION
**Pull request by Electron-libre**
refs #57242

### CODE REVIEW

File: src/librustc_typeck/check/mod.rs [hunk 1]
```diff
@@ -3650,37 +3650,49 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
...
-                    let available_field_names = self.available_field_names(variant);
-                    if !available_field_names.is_empty() {
-                        err.note(&format!("available fields are: {}",
-                                          self.name_series_display(available_field_names)));
+        match variant.ctor_kind {
+            CtorKind::Fn => {
+                err.span_label(field.ident.span, "field does not exist");
+                err.span_label(
+                    field.ident.span,
+                    format!("`{adt}` is a tuple struct, use the appropriate syntax: `{adt}()`",
```
Comment (petrochenkov) [inline src/librustc_typeck/check/mod.rs]:
```suggestion
                    format!("`{adt}` is a tuple {kind_name}, use the appropriate syntax: `{adt}(/* fields */)`",
```

**Comment by petrochenkov** [comment]
LGTM.
r? @estebank on whether `/* fields */` is good enough or should be replaced with an actual fields list.

**Comment by Electron-libre** [comment]
IMHO providing the constructor signature would be a better hint.
Eg: 
```use function constructor MyFoo(u32, String)```

Since only type and order matter for "tuple" syntax.

**Comment by estebank** [comment]
The force push included  src/tools/miri and src/tools/clippy. (It shouldn't.)

**Comment by estebank** [comment]
```
src/librustc_typeck/check/expr.rs:1223: line longer than 100 chars
```

**Review by estebank — CHANGES_REQUESTED** [review]
Ideally we would try to use a structured suggestion that gives out the _correct_ code (which is why @petrochenkov was asking for my input), but I think this is a good first pass at a solution which can be improved upon down the line.

Left some nitpicks inline that shouldn't be too hard to address.

**Comment by estebank** [comment]
> IMHO providing the constructor signature would be a better hint.

In order to do so you would need to modify this code [to not emit one error per field](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=dc3d624c29ac4594978de5dda36dedda), but rather per invocation, and only once you figure out you have the correct amount of fields (and the correct types for them!) you can emit one error with the structured suggestion. I don't want _perfect_ be enemy of the _good enough that will make things materially better_ :)

File: src/librustc_typeck/check/expr.rs [hunk 2]
```diff
@@ -1215,38 +1215,49 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
-                    let available_field_names = self.available_field_names(variant);
-                    if !available_field_names.is_empty() {
-                        err.note(&format!("available fields are: {}",
-                                          self.name_series_display(available_field_names)));
+        match variant.ctor_kind {
+            CtorKind::Fn => {
+                err.span_label(field.ident.span, "field does not exist");
+                err.span_label(
+                    field.ident.span,
+                    format!("`{adt}` is a tuple {kind_name}, use the appropriate syntax: `{adt}(/* fields */)`", adt=ty, kind_name=kind_name)
```
[the remark this answers is not in the document]
Comment (Electron-libre) [reply src/librustc_typeck/check/expr.rs]: done

File: src/librustc_typeck/check/expr.rs [hunk 3]
```diff
@@ -1215,38 +1215,49 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+                );
+            }
+            _ => {
+                // prevent all specified fields from being suggested
+                let skip_fields = skip_fields.iter().map(|ref x| x.ident.as_str());
+                if let Some(field_name) = Self::suggest_field_name(variant,
+                                                                   &field.ident.as_str(),
+                                                                   skip_fields.collect()) {
```
[the remark this answers is not in the document]
Comment (Electron-libre) [reply src/librustc_typeck/check/expr.rs]: done

File: src/librustc_typeck/check/expr.rs [hunk 4]
```diff
@@ -1215,38 +1215,49 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+                        field_name.to_string(),
+                        Applicability::MaybeIncorrect,
+                    );
+                } else {
+                    match ty.sty {
+                        ty::Adt(adt, ..) => {
+                            if adt.is_enum() {
+                                err.span_label(field.ident.span,
+                                               format!("`{}::{}` does not have this field",
+                                                       ty, variant.ident));
```
[the remark this answers is not in the document]
Comment (Electron-libre) [reply src/librustc_typeck/check/expr.rs]: done

File: src/test/ui/numeric/numeric-fields.stderr [hunk 5]
```diff
@@ -2,9 +2,10 @@ error[E0560]: struct `S` has no field named `0b1`
   --> $DIR/numeric-fields.rs:4:15
    |
 LL |     let s = S{0b1: 10, 0: 11};
-   |               ^^^ `S` does not have this field
-   |
-   = note: available fields are: `0`, `1`
+   |               ^^^
+   |               |
+   |               field does not exist
+   |               `S` is a tuple struct, use the appropriate syntax: `S(/* fields */)`
```
[the remark this answers is not in the document]
Comment (Electron-libre) [reply src/test/ui/numeric/numeric-fields.stderr]: done

File: src/librustc_typeck/check/expr.rs [hunk 6]
```diff
@@ -1215,38 +1216,58 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
-                    let available_field_names = self.available_field_names(variant);
-                    if !available_field_names.is_empty() {
-                        err.note(&format!("available fields are: {}",
-                                          self.name_series_display(available_field_names)));
+        match variant.ctor_kind {
+            CtorKind::Fn => {
+                err.span_label(variant.ident.span, format!("`{adt}` defined here", adt=ty));
+                err.span_label(field.ident.span, "field does not exist");
+                err.span_label(ty_span, format!(
+                        "`{adt}` is a tuple {kind_name},\
```
Comment (estebank) [inline src/librustc_typeck/check/expr.rs]: There's a space missing after the comma. You can see it more clearly in the street output.

File: src/librustc_typeck/check/expr.rs [hunk 7]
```diff
@@ -1215,38 +1216,58 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+                        _ => bug!("non-ADT passed to report_unknown_field")
                     }
-                }
-                _ => bug!("non-ADT passed to report_unknown_field")
+                };
             }
-        };
-        err.emit();
+        }
+            err.emit();
```
Comment (estebank) [inline src/librustc_typeck/check/expr.rs]: Indent `err.emit()` aligned to the scope it is in.

**Review by estebank — APPROVED** [review]
R=me once the nitpicks have been addressed.

**Pull request merged**